File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/api/graphql/dataloader/CommentBatchLoader.java
package com.github.arhor.aws.graphql.federation.comments.api.graphql.dataloader;

import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.Comment;
import lombok.RequiredArgsConstructor;
import org.dataloader.MappedBatchLoader;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.function.Function;

@RequiredArgsConstructor
abstract class CommentBatchLoader<T> implements MappedBatchLoader<Long, List<Comment>> {

    private final Executor executor;
    private final Function<Set<Long>, Map<Long, List<Comment>>> function;

    @Override
    public CompletableFuture<Map<Long, List<Comment>>> load(final Set<Long> keys) {
        return CompletableFuture.supplyAsync(() -> function.apply(keys), executor);
    }
}

File: dgs-comments/src/main/kotlin/com/github/arhor/dgs/comments/api/graphql/loader/CommentBatchLoader.kt
package com.github.arhor.dgs.comments.api.graphql.loader

import com.github.arhor.dgs.comments.generated.graphql.types.Comment
import com.github.arhor.dgs.comments.service.CommentService
import com.netflix.graphql.dgs.DgsDataLoader
import org.dataloader.MappedBatchLoader
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.function.Function

sealed class CommentBatchLoader(
    private val executor: Executor,
    private val loadingFunction: Function<Set<Long>, Map<Long, List<Comment>>>,
) : MappedBatchLoader<Long, List<Comment>> {

    override fun load(keys: Set<Long>): CompletableFuture<Map<Long, List<Comment>>> {
        return CompletableFuture.supplyAsync({ loadingFunction.apply(keys) }, executor)
    }

    @DgsDataLoader(maxBatchSize = 50)
    class ForUser(asyncExecutor: Executor, commentService: CommentService) : CommentBatchLoader(
        executor = asyncExecutor,
        loadingFunction = commentService::getCommentsByUserIds
    )

    @DgsDataLoader(maxBatchSize = 50)
    class ForPost(asyncExecutor: Executor, commentService: CommentService) : CommentBatchLoader(
        executor = asyncExecutor,
        loadingFunction = commentService::getCommentsByPostIds
    )
}
